File: env_store.py
from __future__ import annotations

from pathlib import Path
from typing import Mapping


BACKUP_SUFFIX = ".bak"
TEMP_SUFFIX = ".tmp"


def update_env_file(env_file: str, updates: Mapping[str, str]) -> None:
    """.env を安全に更新する。"""
    path = Path(env_file)
    path.parent.mkdir(parents=True, exist_ok=True)

    backup_path = path.with_suffix(path.suffix + BACKUP_SUFFIX)
    original_text = ""
    if path.exists():
        original_text = path.read_text(encoding="utf-8")
        if not original_text.strip() and backup_path.exists():
            original_text = backup_path.read_text(encoding="utf-8")

    lines = original_text.splitlines()
    updated_keys = set()
    new_lines: list[str] = []

    for line in lines:
        stripped = line.strip()
        if not stripped or stripped.startswith("#") or "=" not in line:
            new_lines.append(line)
            continue
        key, _value = line.split("=", 1)
        key = key.strip()
        if key in updates:
            new_lines.append(f"{key}={updates[key]}")
            updated_keys.add(key)
        else:
            new_lines.append(line)

    for key, value in updates.items():
        if key not in updated_keys:
            new_lines.append(f"{key}={value}")

    new_text = "\n".join(new_lines)
    if new_text:
        new_text += "\n"

    if path.exists():
        backup_path.write_text(original_text, encoding="utf-8")

    temp_path = path.with_suffix(path.suffix + TEMP_SUFFIX)
    temp_path.write_text(new_text, encoding="utf-8")
    temp_path.replace(path)

File: src/utils/env-store.ts
import fs from "fs";
import path from "path";

const BACKUP_SUFFIX = ".backup";
const TEMP_SUFFIX = ".tmp";

/**
 * .envファイルを安全に更新する（コメントと空行を保持）
 */
export function updateEnvFile(
  envFile: string,
  updates: Record<string, string>
): void {
  const absPath = path.resolve(envFile);
  const backupPath = absPath + BACKUP_SUFFIX;
  const tempPath = absPath + TEMP_SUFFIX;

  let lines: string[] = [];
  if (fs.existsSync(absPath)) {
    lines = fs.readFileSync(absPath, "utf-8").split("\n");
  }

  const remaining = new Set(Object.keys(updates));
  const newLines: string[] = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed === "" || trimmed.startsWith("#")) {
      newLines.push(line);
      continue;
    }

    const eqIdx = trimmed.indexOf("=");
    if (eqIdx === -1) {
      newLines.push(line);
      continue;
    }

    const key = trimmed.substring(0, eqIdx).trim();
    if (key in updates) {
      newLines.push(`${key}=${updates[key]}`);
      remaining.delete(key);
    } else {
      newLines.push(line);
    }
  }

  // 新しいキーを追加
  for (const key of remaining) {
    newLines.push(`${key}=${updates[key]}`);
  }

  // バックアップ作成
  if (fs.existsSync(absPath)) {
    fs.copyFileSync(absPath, backupPath);
  }

  // 一時ファイル経由でアトミック書き込み
  fs.writeFileSync(tempPath, newLines.join("\n"), "utf-8");
  fs.renameSync(tempPath, absPath);
}
